mqtrigger/messageQueue/nats.go

```diff
@@ -0,0 +1,182 @@
+/*
+Copyright 2016 The Fission Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package messageQueue
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	ns "github.com/nats-io/go-nats-streaming"
+	nsUtil "github.com/nats-io/nats-streaming-server/util"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/fission/fission"
+	"github.com/fission/fission/publisher"
+)
+
+const (
+	natsClusterID  = "fissionMQTrigger"
+	natsProtocol   = "nats://"
+	natsClientID   = "fission"
+	natsQueueGroup = "fission-messageQueueNatsTrigger"
+)
+
+type (
+	Nats struct {
+		MessageQueueTriggerManager
+		nsConn      *ns.Conn
+		nsPublisher *publisher.NatsPublisher
+	}
+)
+
+// To enable nats authentication, please follow the instruction described in
+// http://nats.io/documentation/server/gnatsd-authentication/.
+// And dont forget to change the MESSAGE_QUEUE_URL in mqtrigger deployment.
+
+func makeNatsTriggerManager(mqTriggerMgr MessageQueueTriggerManager) (MessageQueueTriggerManagerInterface, error) {
+	conn, err := ns.Connect(natsClusterID, natsClientID, ns.NatsURL(mqTriggerMgr.mqCfg.Url))
+	if err != nil {
+		return nil, err
+	}
+	nsPublisher := publisher.MakeNatsPublisher(&conn)
+	nats := Nats{
+		MessageQueueTriggerManager: mqTriggerMgr,
+		nsConn:      &conn,
+		nsPublisher: nsPublisher,
+	}
+	go nats.sync()
+	return &nats, nil
+}
+
+func (nats *Nats) add(trigger fission.MessageQueueTrigger) error {
+	subj := trigger.Topic
+
+	if !isTopicValidForNats(subj) {
+		return errors.New(fmt.Sprintf("Not a valid topic: %s", trigger.Topic))
+	}
+
+	handler := func(msg *ns.Msg) {
+		url := nats.routerUrl + "/" + strings.TrimPrefix(fission.UrlForFunction(&trigger.Function), "/")
+		log.Printf("Making HTTP request to %v", url)
+
+		// Create request
+		resp, err := http.Post(url, "", bytes.NewReader(msg.Data))
+		if err != nil {
+			log.Warningf("Request failed: %v", url)
+			return
+		}
+		defer resp.Body.Close()
+
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			log.Warningf("Request body error: %v", string(body))
+			return
+		}
+		if resp.StatusCode != 200 {
+			log.Printf("Request returned failure: %v", resp.StatusCode)
+			return
+		}
+		// trigger acks message only if a request done successfully
+		err = msg.Ack()
+		if err != nil {
+			log.Warningf("Failed to ack message: %v", err)
+		}
+		if len(trigger.ResponseTopic) > 0 {
```

Comment:
I think we should keep the response topic for the following reasons:
1. Nats-Streaming does not provide the request/reply as Nats does. (https://godoc.org/github.com/nats-io/go-nats-streaming#Msg)
2. For the other message queue service, they may not provide the reply topic such as CloudPubSub (https://godoc.org/cloud.google.com/go/pubsub#Message).